src/main/java/dev/morling/onebrc/CalculateAverage_makohn.java

```diff
@@ -0,0 +1,287 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+//
+// This implementation is partially inspired by
+//
+// - GavinRay97: 1BRC in Kotlin (memory mapping, chunking) | https://github.com/gunnarmorling/1brc/discussions/154
+// - dannyvankooten: 1BRC in C (integer parsing, linear probing) | https://github.com/gunnarmorling/1brc/discussions/46
+//
+public class CalculateAverage_makohn {
+
+    private static final String FILE = "./measurements.txt";
+
+    private static class Measurement implements Comparable<Measurement> {
+        final String city;
+        int min;
+        int max;
+        int count = 1;
+        int sum;
+
+        Measurement(String city, int val) {
+            this.city = city;
+            this.min = val;
+            this.max = val;
+            this.sum = val;
+        }
+
+        @Override
+        public String toString() {
+            return STR."\{city}=\{round(min)}/\{round((1.0 * sum) / count)}/\{round(max)}";
+        }
+
+        private double round(double value) {
+            return Math.round(value) / 10.0;
+        }
+
+        @Override
+        public int compareTo(Measurement other) {
+            return this.city.compareTo(other.city);
+        }
+    }
+
+    // Convert a given byte array of temperature data to an int value
+    // Since the temperate values only have one decimal, we can use integer arithmetic until the end
+    //
+    // buffer: [..., '-', '1', '9', '.', '7', ...]
+    // -------------> offset
+    // ............ = s
+    //
+    // We initialize a "pointer" s with the offset. Depending on whether the first char is a '-' or not, we set the
+    // sign and increment the pointer.
+    //
+    // Then we only have to distinguish between one-digit and two-digit numbers.
+    // Depending on that, we set an index for the respective parts of the number.
+    //
+    private static int toInt(byte[] in, int offset) {
+        int sign = 1;
+        int s = offset;
+        if (in[s] == '-') {
+            sign = -1;
+            s++;
+        }
+
+        if (in[s + 1] == '.')
+            return sign * ((in[s] - '0') * 10 + (in[s + 2] - '0'));
+
+        return sign * ((in[s] - '0') * 100 + (in[s + 1] - '0') * 10 + (in[s + 3] - '0'));
+    }
+
+    // 10_000 distinct station names as per specification
+    // We use the next power of two (2^14 = 16384) to allow for bit-masking our hash (instead of using modulo)
+    private static final int MAX_STATIONS = 2 << 14;
+
+    // Twice as big as the maximum number of stations
+    private static final int MAP_CAPACITY = MAX_STATIONS * 2;
+
+    // We start at 1 to allow for checking our hash-index map for > 0
+    private static final int RES_FIRST_INDEX = 1;
+
+    private static class ResultMap {
+        final int[] map = new int[MAP_CAPACITY]; // hash -> index
+        final Measurement[] measurements = new Measurement[MAX_STATIONS]; // index -> measurement
```

Comment:
I've seen this a couple of times, what's the advantage of having this indirection with two arrays, rather than probing within one?

Reply:
That way, it's easier/faster to iterate over the `Measurement` array, since its entries are stored in order without gaps in between (and we know the number of actual entries as per `lastIndex`) (see lines 261 and 280)

If you would use the hash directly as an index, you need to filter out null/0 values later when merging the results or when building the output string.

Though, there might be some clever way to design a hash function to reduce/avoid the gaps, achieving a more dense distribution.

Reply:
I see, yes, it can save you some iterations. Makes sense, thx!